### PR TITLE
tests/: add DM CASE 104~112

### PR DESCRIPTION
**Pull request by lichunzhu**
### What problem does this PR solve? 
Add more test cases for DM optimistic shard ddl.

### What is changed and how it works?
1. DM-104: add not null columns with default value.
2. DM-105: add indices.
3. DM-106: drop indices.
4. DM-107: add not null columns without default value.
5. DM-108: add columns with different decimal.
6. DM-109: add columns with different fieldLen(newly added column has a larger flen).
7. DM-110: add columns with different fieldLen(newly added column has a smaller flen)..
8. DM-111: add columns with different zerofill.
8. DM-112: add columns with different unsigned.

### Check List 

Tests 

 - Unit test
 - Integration test

Related changes

 - Need to cherry-pick to the release branch

### CODE REVIEW

File: tests/shardddl3/run.sh [hunk 1]
```diff
@@ -679,6 +679,197 @@ function DM_103() {
...
+    run_sql_tidb_with_retry "select count(1) from ${shardddl}.${tb};" "count(1): 9"
+}
+
+function DM_104() {
+    # currently not support pessimistic
+    # run_case 104 "double-source-pessimistic" "init_table 111 211 212" "clean_table" "pessimistic"
+    run_case 104 "double-source-optimistic" "init_table 111 211 212" "clean_table" "optimistic"
+}
+
+function ADD_DROP_INDEX_CASE() {
```
Comment (lance6716) [inline tests/shardddl3/run.sh:707]: hope the helper function has a lower case name or something, to distinguish it from DM cases

File: pkg/shardddl/optimism/lock.go [hunk 2]
```diff
@@ -214,6 +216,13 @@ func (l *Lock) TrySync(callerSource, callerSchema, callerTable string,
 			// for these two cases, we should execute the DDLs to the downstream to update the schema.
 			log.L().Info("joined table info changed", zap.String("lock", l.ID), zap.Int("cmp", cmp), zap.Stringer("from", oldJoined), zap.Stringer("to", newJoined),
 				zap.String("source", callerSource), zap.String("schema", callerSchema), zap.String("table", callerTable), zap.Strings("ddls", ddls))
+			// check for add column with different field lengths
+			if cmp < 0 {
+				err = AddDifferentFieldLenColumns(l.ID, ddls[idx], oldJoined, newJoined)
```
[the remark this answers is not in the document]
Comment (lichunzhu) [reply pkg/shardddl/optimism/lock.go:221]: This one is for add columns check.

**Pull request merged**